test/unit/app/utils/ussd/test_token_processor.py

```diff
@@ -0,0 +1,147 @@
+    user = UserFactory(preferred_language=preferred_language, phone=phone())
+    if user_type == "group":
+        user.set_held_role('GROUP_ACCOUNT', 'grassroots_group_account')
+
+    token1 = Token.query.filter_by(symbol="SM1").first()
+    create_transfer_account_for_user(user, token1, 200)
+
+    def mock_convert(exchange_contract, from_token, to_token, from_amount):
+            return from_amount * 1.2
+    mocker.patch('server.bt.get_conversion_amount', mock_convert)
+
+    def mock_send_message(phone, message):
+        assert exchange_text in message
+        assert limit_text in message
+    mocker.patch('server.message_processor.send_message', mock_send_message)
+    TokenProcessor.fetch_exchange_rate(user)
+
+
+def test_send_token(mocker, test_client, init_database, initialised_blockchain_network):
+    sender = UserFactory(preferred_language="en", phone=phone(), first_name="Bob", last_name="Foo")
+    token1 = Token.query.filter_by(symbol="SM1").first()
+    create_transfer_account_for_user(sender, token1, 200)
+
+    recipient = UserFactory(phone=phone(), first_name="Joe", last_name="Bar")
+    token2 = Token.query.filter_by(symbol="SM2").first()
+    create_transfer_account_for_user(recipient, token2, 300)
+
+    def mock_convert(exchange_contract, from_token, to_token, from_amount, signing_address):
+        if from_token.symbol == "SM1":
+            return from_amount * 1.5
+        else:
+            return from_amount * 0.75
+    mocker.patch('server.bt.get_conversion_amount', mock_convert)
+
+    messages = []
+
+    def mock_send_message(phone, message):
+        messages.append({'phone': phone, 'message': message})
+    mocker.patch('server.message_processor.send_message', mock_send_message)
+
+    TokenProcessor.send_token(sender, recipient, 10, "A reason", 1)
+    assert default_transfer_account(sender).balance == 190
+    # TODO: shouldn't it double convert from the reserve to be 320..?
+    assert default_transfer_account(recipient).balance == 315
+
+    assert len(messages) == 2
+    sent_message = messages[0]
+    assert sent_message['phone'] == sender.phone
+    assert 'sent a payment of 10 SM1 = 15.0 SM2' in sent_message['message']
+    received_message = messages[1]
+    assert received_message['phone'] == recipient.phone
+    assert 'received a payment of 15.0 SM2 = 10 SM1' in received_message['message']
+
+
+def test_exchange_token(mocker, test_client, init_database, initialised_blockchain_network):
+    sender = UserFactory(preferred_language="en", phone=phone(), first_name="Bob", last_name="Foo")
+    token1 = Token.query.filter_by(symbol="SM1").first()
+    create_transfer_account_for_user(sender, token1, 200)
+
+    agent = UserFactory(phone=phone(), first_name="Joe", last_name="Bar")
+    agent.set_held_role('TOKEN_AGENT', 'grassroots_token_agent')
+    # this is under the assumption that token agent would have default token being the reserve token. is this the case?
```

Comment:
do we know if this is a valid assumption?

Reply:
I don't think this is a valid assumption right now, but it _should be_. I'll create a todo for 'review token agent setup'